src/core/env.ml

```diff
@@ -1,79 +1,98 @@
-(** Scoping environment for variables. *)
+(** Scoping environment for variables.
+    Note that since meta-variable types are closed, the contexts passed to
+    {!val:Eval.whnf} and alike are empty. *)
 
+open Extra
 open Terms
 
 (** Type of an environment, used in scoping to associate names to
-   corresponding Bindlib variables and types. Note that it cannot be
-   implemented by a map as the order is important. *)
-type env = (string * (tvar * tbox)) list
+    corresponding Bindlib variables and types. Note that it cannot be
+    implemented by a map as the order is important. The boolean is used to
+    indicate whether the variable is defined in some context. *)
+type env = (string * (tvar * bool * tbox)) list
 
 type t = env
 
-(** [empty] is the empty environemnt. *)
+(** [empty] is the empty environment. *)
 let empty : env = []
 
-(** [add v a env] extends the environment [env] by mapping the string
-   [Bindlib.name_of v] to [(v,a)]. *)
-let add : tvar -> tbox -> env -> env = fun v a env ->
-  (Bindlib.name_of v, (v, a)) :: env
+(** [add v d a env] extends the environment [env] by mapping the string
+    [Bindlib.name_of v] to [(v,d,a)]. *)
+let add : tvar -> bool -> tbox -> env -> env = fun v d a env ->
+  (Bindlib.name_of v, (v, d, a)) :: env
 
 (** [find n env] returns the Bindlib variable associated to the variable  name
     [n] in the environment [env]. If none is found, [Not_found] is raised. *)
 let find : string -> env -> tvar = fun n env ->
-  fst (List.assoc n env)
+  let (x,_,_) = (List.assoc n env) in x
 
-(** [to_prod env t] builds a sequence of products whose  domains  are the
-    variables of the environment [env] (from left to right), and which body is
-    the term [t]: [to_prod [(xn,an);..;(x1,a1)] t = ∀x1:a1,..,∀xn:an,t]. *)
+(** [to_prod env t] builds a sequence of products whose domains are the {e not
+    defined} variables of the environment [env] (from left to right), and
+    which body is the term [t]: [to_prod [(xn,an);..;(x1,a1)] t =
+    ∀x1:a1,..,∀xn:an,t]. *)
 let to_prod : env -> tbox -> term = fun env t ->
-  let fn t (_,(x,a)) = _Prod a (Bindlib.bind_var x t) in
+  let fn t (_,(x,d,a)) = if d then t else _Prod a (Bindlib.bind_var x t) in
```

Comment:
Ah, but we cannot simply ignore these defined variables like this I guess. We need to replace them by let's:
the environment [x:A:=a,y:B] applied to u should generate the term [let x:A:=a in ∀y:B,u].
So we need to have the definitions as well in the type env!